libs/ngx-nuts-and-bolts/src/lib/utilities/table-state/table-state.spec.ts

```diff
@@ -0,0 +1,227 @@
+		};
+
+		expect(filtersValue).toBe(null);
+
+		await changeFilters(router, filter);
+
+		filtersValue = await createCustomFiltersObservable(route, router).pipe(take(1)).toPromise();
+		expect(navigationSpy).toHaveBeenCalledTimes(1);
+		expect(filtersValue).not.toBe(null);
+	});
+
+	it('should set custom filter observable value to null if filter is present but all options are empty', async () => {
+		const callbacks = mockSubscribeCallbacks();
+
+		createCustomFiltersObservable(route, router).subscribe(callbacks);
+		expect(callbacks.next).toHaveBeenLastCalledWith(null);
+
+		const filters = {
+			arrayQuery: [],
+			textQuery: '',
+			numberQuery: undefined,
+			objectQuery: undefined,
+		};
+
+		await changeFilters(router, filters);
+		expect(callbacks.next).toHaveBeenLastCalledWith(null);
+	});
+
+	it('should set custom filter observable value to null if array filter is empty', async () => {
+		const callbacks = mockSubscribeCallbacks();
+
+		createCustomFiltersObservable(route, router).subscribe(callbacks);
+		expect(callbacks.next).toHaveBeenLastCalledWith(null);
+
+		const filters = {
+			arrayQuery: [],
+		};
+
+		await changeFilters(router, filters);
+		expect(callbacks.next).toHaveBeenLastCalledWith(null);
+	});
+
+	it('should set custom filter observable value to null if text filter is empty', async () => {
+		const callbacks = mockSubscribeCallbacks();
+		createCustomFiltersObservable(route, router).subscribe(callbacks);
+		expect(callbacks.next).toHaveBeenLastCalledWith(null);
+
+		const filters = {
+			textQuery: '',
+		};
+
+		await changeFilters(router, filters);
+		expect(callbacks.next).toHaveBeenLastCalledWith(null);
+	});
+
+	it('should set custom filter observable value to null if number filter is empty', async () => {
+		const callbacks = mockSubscribeCallbacks();
+
+		createCustomFiltersObservable(route, router).subscribe(callbacks);
+		expect(callbacks.next).toHaveBeenLastCalledWith(null);
+
+		const filters = {
+			numberQuery: undefined,
+		};
+
+		await changeFilters(router, filters);
+		expect(callbacks.next).toHaveBeenLastCalledWith(null);
+	});
+
+	it('should not set custom filter observable value to null if number filter is 0', async () => {
+		const callbacks = mockSubscribeCallbacks();
+
+		createCustomFiltersObservable(route, router).subscribe(callbacks);
+		expect(callbacks.next).toHaveBeenLastCalledWith(null);
+
+		const filters = {
+			numberQuery: 0,
+		};
+
+		await changeFilters(router, filters);
+		expect(callbacks.next).not.toHaveBeenLastCalledWith(null);
+	});
+
+	it('should set custom filter observable value to null if object filter is empty', async () => {
+		const callbacks = mockSubscribeCallbacks();
+
+		createCustomFiltersObservable(route, router).subscribe(callbacks);
+		expect(callbacks.next).toHaveBeenLastCalledWith(null);
+
+		const filters = {
+			objectQuery: undefined,
+		};
+
+		await changeFilters(router, filters);
+		expect(callbacks.next.mock.calls.pop()).not.toBe(null);
```

Comment:
Please don't use pop, as you've seen it can cause some hard to find errors with `toHaveBeenCalledTimes`. Same comment for line 221